src/Data/Machine/Group.hs

```diff
@@ -1,59 +1,162 @@
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.State.Strict (get, put)
 import Data.Machine
+import Data.Machine.Lift (execStateM)
+
+#if __GLASGOW_HASKELL__ < 710
+-- $setup
+-- >>> import Control.Monad.Trans.Reader (ask, runReader)
+-- >>> import Control.Applicative ((<$))
+#else
+-- $setup
+-- >>> import Control.Monad.Trans.Reader (ask, runReader)
+#endif
 
 isLeft :: Either a b -> Bool
 isLeft = either (const True) (const False)
 
--- | Using a function to signal group changes, apply a machine independently over each group.
-groupingOn :: Monad m => (a -> a -> Bool) -> ProcessT m a b -> ProcessT m a b
-groupingOn f m = taggedBy f ~> partitioning m
+-- | Using a function to signal group changes, apply a machine independently
+-- over each group.
+groupingOn_ :: Monad m => (a -> a -> Bool) -> ProcessT m a b -> ProcessT m a b
+groupingOn_ f m = taggedBy_ f ~> partitioning_ m
+{-# INLINE groupingOn_ #-}
+
+-- | Using a function to signal group changes, apply a machine independently
+-- over each group using the difference between the group edges.
+groupingOn :: Monad m => i -> (a -> a -> Maybe i) -> (i -> ProcessT m a b) -> ProcessT m a b
+groupingOn i0 f m = taggedBy f ~> partitioning i0 m
+{-# INLINE groupingOn #-}
 
 -- | Mark a transition point between two groups as a function of adjacent elements.
 -- Examples
 --
--- >>> runT $ supply [1,2,2] (taggedBy (==))
+-- >>> runT $ supply [1,2,2] (taggedBy_ (==))
 -- [Right 1,Left (),Right 2,Right 2]
-taggedBy :: Monad m => (a -> a -> Bool) -> ProcessT m a (Either () a)
+taggedBy_ :: Monad m => (a -> a -> Bool) -> ProcessT m a (Either () a)
+taggedBy_ f = taggedBy (\x y -> guard (not (f x y)))
+{-# INLINE taggedBy_ #-}
+
+-- | Mark a transition point between two groups as a function of adjacent elements,
+-- and let their difference be the separator.
+-- Examples
+--
+-- >>> runT $ supply [1,3,3,2] (taggedBy (\x y -> (y > x) <$ guard (x /= y)))
+-- [Right 1,Left True,Right 3,Right 3,Left False,Right 2]
+taggedBy :: Monad m => (a -> a -> Maybe i) -> ProcessT m a (Either i a)
 taggedBy f = construct $ await >>= go
-  where go x = do
-          yield (Right x)
-          y <- await
-          if not (f x y) then yield (Left ()) >> go y else go y
+  where
+    go x = do
+      yield (Right x)
+      y <- await
+      maybe (go y) (\z -> yield (Left z) >> go y) (f x y)
+
+-- | Mark a transition point between two groups at every 'n' values.
+-- Examples
+--
+-- >>> runT $ supply [1..5] (taggedAt_ 2)
+-- [Right 1,Right 2,Left (),Right 3,Right 4,Left (),Right 5]
+taggedAt_ :: Monad m => Int -> ProcessT m a (Either () a)
+taggedAt_ n = taggedAt n () id
+{-# INLINE taggedAt_ #-}
 
+-- | Mark a transition point between two groups at every 'n' values, using the
+-- counter as the separator.
+-- Examples
+--
+-- >>> runT $ supply [1..5] (taggedCount 2 1)
+-- [Right 1,Right 2,Left 1,Right 3,Right 4,Left 2,Right 5]
+taggedCount :: (Enum i, Monad m) => Int -> i -> ProcessT m a (Either i a)
```

Comment:
Enum is really lousy. Do we really need that, or can we make do with `Int`? I would think users could convert to whatever they like by mapping over the `ProcessT`.

Reply:
Yeah, that's probably fine.